tests/python_tests/models/test_couplings.py

```diff
@@ -0,0 +1,590 @@
+
+
+# TEST SPIN COUPLINGS
+
+
+def test_spin_spin_coupling(any_backend, np_random):
+    check_dense_blocks = np_random.choice([True, False])
+    site_list = generate_spin_dofs(any_backend)
+    num_sites = min(5, len(site_list))
+    site_list = np_random.choice(site_list, size=num_sites, replace=False)
+    for i, site1 in enumerate(site_list):
+        Jx, Jy, Jz = np_random.random(3)
+        # either SpinSite or SpinHalfFermionSite
+        conserve = site1.conserve if isinstance(site1, sites.SpinSite) else site1.conserve_S
+        if conserve in ['Sz']:
+            Jx = Jy = 0
+        elif conserve in ['SU(2)']:
+            Jx = Jy = Jz
+
+        # test different site combinations
+        for site2 in site_list[:i + 1]:
+            # Note: is_same_symmetry does not work here since it does not distinguish
+            # between U(1) fermion number symmetry and Sz spin symmetry for fermions
+            if not site1.symmetry == site2.symmetry:
+                continue
+            coupling = couplings.spin_spin_coupling([site1, site2], Jx=Jx, Jy=Jy, Jz=Jz)
+            coupling.test_sanity()
+            if check_dense_blocks:
+                expect = site1.spin_vector.copy()
+                expect[:, :, 0] *= Jx
+                expect[:, :, 1] *= Jy
+                expect[:, :, 2] *= Jz
+                expect = np.tensordot(expect, site2.spin_vector, axes=[2, 2])
+                expect = np.transpose(expect, [0, 2, 3, 1])
+                assert np.allclose(coupling.to_numpy(understood_braiding=True), expect)
+
+    # test correct number of sites
+    with pytest.raises(AssertionError):
+        _ = couplings.spin_spin_coupling([site_list[0]], Jx=1., Jy=1., Jz=1.)
+    with pytest.raises(AssertionError):
+        _ = couplings.spin_spin_coupling([site_list[0]] * 3, Jx=1., Jy=1., Jz=1.)
+
+
+def test_spin_field_coupling(any_backend, np_random):
+    check_dense_blocks = np_random.choice([True, False])
+    site_list = generate_spin_dofs(any_backend)
+    num_sites = min(5, len(site_list))
+    site_list = np_random.choice(site_list, size=num_sites, replace=False)
+    for site in site_list:
+        hx, hy, hz = np_random.random(3)
+        # either SpinSite or SpinHalfFermionSite
+        conserve = site.conserve if isinstance(site, sites.SpinSite) else site.conserve_S
+        if conserve in ['Sz', 'parity']:
+            hx = hy = 0
+        elif conserve in ['SU(2)']:
+            # coupling not allowed
+            continue
+        coupling = couplings.spin_field_coupling([site], hx=hx, hy=hy, hz=hz)
+        coupling.test_sanity()
+        if check_dense_blocks:
+            expect = site.spin_vector
+            expect = hx * expect[:, :, 0] + hy * expect[:, :, 1] + hz * expect[:, :, 2]
+            assert np.allclose(coupling.to_numpy(understood_braiding=True), expect)
+
+    # test correct number of sites
+    with pytest.raises(AssertionError):
```

Comment:
as a general pattern, if a failure mode is relevant enough that we want to test for it, it should have a proper test type (here `ValueError`) and a useful error message, and we should check for the message here with ``raises(ValueError, match='...')``